src/Workspaces/Core/Portable/Storage/SQLite/v2/Interop/SqlConnection.cs

```diff
@@ -295,6 +324,77 @@ private Stream ReadBlobIntoPooledStream(SafeSqliteBlobHandle blob, int length)
             }
         }
 
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
+        public Optional<T> ReadBlob_MustRunInTransaction<T>(
+            Database database, Table table, Column column, long rowId,
+            Func<SqlConnection, SafeSqliteBlobHandle, Optional<T>> readBlob)
+        {
+            // NOTE: we do need to do the blob reading in a transaction because of the
+            // following: https://www.sqlite.org/c3ref/blob_open.html
+            //
+            // If the row that a BLOB handle points to is modified by an UPDATE, DELETE, 
+            // or by ON CONFLICT side-effects then the BLOB handle is marked as "expired".
+            // This is true if any column of the row is changed, even a column other than
+            // the one the BLOB handle is open on. Calls to sqlite3_blob_read() and 
+            // sqlite3_blob_write() for an expired BLOB handle fail with a return code of
+            // SQLITE_ABORT.
+            if (!IsInTransaction)
+            {
+                throw new InvalidOperationException("Must read blobs within a transaction to prevent corruption!");
+            }
+
+            var databaseNameBytes = database switch
+            {
+                Database.Main => s_mainNameWithTrailingZero,
+                Database.WriteCache => s_writeCacheNameWithTrailingZero,
+                _ => throw ExceptionUtilities.UnexpectedValue(database),
+            };
+
+            var tableNameBytes = table switch
+            {
+                Table.Solution => s_solutionTableName,
+                Table.Project => s_projectTableName,
+                Table.Document => s_documentTableName,
+                _ => throw ExceptionUtilities.UnexpectedValue(table),
+            };
+
+            var columnNameBytes = column switch
+            {
+                Column.Data => s_dataColumnName,
+                Column.Checksum => s_checksumColumnName,
+                _ => throw ExceptionUtilities.UnexpectedValue(column),
+            };
+
+            unsafe
+            {
+                fixed (byte* databaseNamePtr = databaseNameBytes)
+                fixed (byte* tableNamePtr = tableNameBytes)
+                fixed (byte* columnNamePtr = columnNameBytes)
```

Comment:
ick.  but i couldn't figure out any other way to do this given the API exposed by sqlitepcl.  @ericsink any thoughts?

Reply:
@ericsink could you expose an overload that takes a ReadOnlySpan<byte>?  That would let us avoid the use of unsafe here.  If that overload already exists, let me know!

Reply:
Still catching up, I just saw these, but I think we've already addressed them with the PR you sent.